rust/connlib/tunnel/src/peer_store.rs

```diff
@@ -0,0 +1,141 @@
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::net::IpAddr;
+
+use crate::peer::{PacketTransform, Peer};
+use ip_network::IpNetwork;
+use ip_network_table::IpNetworkTable;
+
+pub struct PeerStore<TId, TTransform> {
+    id_by_ip: IpNetworkTable<TId>,
+    peer_by_id: HashMap<TId, Peer<TId, TTransform>>,
+}
```

Comment:
Actually, I think there might be a slightly better design here:

```suggestion
pub struct PeerStore<TId, V> {
    id_by_ip: IpNetworkTable<TId>,
    peer_by_id: HashMap<TId, V>,
}
```

If we make this thing generic over the actual value that it stores, we can split `Peer` by client / gateway too which allows us to get rid of the "transform" abstraction. I think that is useful because `PeerStore` is now used exclusively within `ClientState` and `GatewayState`, meaning there is no need for this "transform" abstraction, e.g. for gateways, `packet_transform` doesn't do anything so we should just be able to inline that call and pass the packet on.

We don't need to split `Peer` within this PR but I think long-term, it would be useful to strive for only having one layer where we spit between client and gateway: The tunnel's state. All other data structures should be agnostic over that separation.

Reply:
This can probably be simplified further if you remove the `conn_id` from `Peer`.